odbc/src/api/data.rs

```diff
@@ -129,28 +130,34 @@ impl IntoCData for Bson {
             // We should update this when the bson crate supports Decimal128 entirely.
             Bson::Decimal128(_) => {
                 let out = self.to_f64()?;
-                Ok((
-                    out as i64,
-                    if out.floor() != out {
-                        Some(ODBCError::FractionalTruncation(out.to_string()))
-                    } else {
-                        None
-                    },
-                ))
+                if out > i64::MAX as f64 || out < i64::MIN as f64 {
+                    Err(ODBCError::IntegralTruncation(out.to_string()))
+                } else {
+                    Ok((
+                        out as i64,
+                        if out.floor() != out {
+                            Some(ODBCError::FractionalTruncation(out.to_string()))
+                        } else {
+                            None
+                        },
+                    ))
+                }
             }
             o => Err(ODBCError::RestrictedDataType(o.to_type_str(), INT64)),
         }
     }
 
     fn to_i32(&self) -> Result<(i32, Option<ODBCError>)> {
         match self {
-            Bson::Double(x) if *x > i32::MAX as f64 => {
+            Bson::Double(x) if *x > i32::MAX as f64 || *x < i32::MIN as f64 => {
                 Err(ODBCError::IntegralTruncation(x.to_string()))
             }
-            Bson::Int64(x) if *x > i32::MAX as i64 => {
+            Bson::Int64(x) if *x > i32::MAX as i64 || *x < i32::MIN as i64 => {
                 Err(ODBCError::IntegralTruncation(x.to_string()))
             }
-            Bson::Decimal128(x) if self.to_f64()? > i32::MAX as f64 => {
+            Bson::Decimal128(x)
+                if self.to_f64()? > i32::MAX as f64 || self.to_f64()? < i32::MIN as f64 =>
+            {
                 Err(ODBCError::IntegralTruncation(x.to_string()))
             }
             _ => self.to_i64().map_or_else(
```

Comment:
umh, I am wondering if this is correct for string to i32 if the string represent a value outside of the i32 boundaries? Similarly to the case above `Bson::Int64(x) if *x > i32::MAX as i64 || *x < i32::MIN as i64`

Reply:
I don't like how decimals are checked and plan to revisit them when the Rust DBX team implements the type entirely to do proper bounds checking.